Studio CLI: Add bundled Node binary and fix Windows popup

bcotrim opened the pull request:
## Related issues

- Fixes STU-1148

## Proposed Changes

1. **Add bundled Node.js binary** - The CLI now uses a bundled Node.js binary instead of relying on the Electron run as Node.

2. **Fix Windows terminal popup** - Patched the `ps-man` package to add `windowsHide: true` to spawn calls. This prevents a brief console window popup (showing `C:\WINDOWS\system32\tasklist`) when running CLI commands on Windows.

## Testing Instructions

### Windows
1. Build the CLI: `npm run cli:build`
2. Run `node dist/cli/main.js site create --name=test-site`
3. Verify no popup windows appear during site creation
4. Verify the site is created successfully
5. Create a site using Studio and confirm operation times are similar

### macOS
1. Build the CLI: `npm run cli:build`  
2. Run `node dist/cli/main.js site create --name=test-site`
3. Verify the site is created successfully
4. Create a site using Studio and confirm operation times are similar

## Pre-merge Checklist

- [x] Have you checked for TypeScript, React or other console errors?

bcotrim committed:
add bundled node

fredrikekelund commented on scripts/download-node-binary.sh:
To make this more portable, let's do it in JS instead.

bcotrim committed:
change node download script

bcotrim committed:
Fix cli/node_modules copy pattern for Windows compatibility

bcotrim committed:
fix node command on Windows

bcotrim committed:
hide windows on child processes

bcotrim committed:
ensure nvrm version is used

bcotrim committed:
logs

bcotrim committed:
fix lint and tests

bcotrim committed:
attempt to fix windows hide

bcotrim committed:
trigger tests

nightnei commented on scripts/download-node-binary.js:
```diff
@@ -0,0 +1,189 @@
+	const nvmrcPath = path.join( __dirname, '..', '.nvmrc' );
+	if ( fs.existsSync( nvmrcPath ) ) {
+		const version = fs.readFileSync( nvmrcPath, 'utf-8' ).trim();
+		// Ensure version starts with 'v'
```
The code looks self-explanatory
```suggestion
```

nightnei commented on src/storage/paths.ts:
```diff
@@ -66,6 +66,19 @@ export function getCliPath(): string {
+export function getBundledNodeBinaryPath(): string {
+	const nodeBinaryName = process.platform === 'win32' ? 'node.exe' : 'node';
+
+	if ( process.env.NODE_ENV === 'development' || process.env.NODE_ENV === 'test' ) {
```
I am thinking - should we keep it only for tests? To have it used in development, to work with the same as in prod, to avoid if in prod something will be unexpectedly broken?
```suggestion
	if ( process.env.NODE_ENV === 'test' ) {
```

fredrikekelund replied:
The only time we can guarantee that the bundled node binary is available is in the production build. I flipped the logic in this function to return the bundled path only when `NODE_ENV === 'production'`

nightnei reviewed (approved):
The code looks great and works as expected:
1. I don't see popup on windows anymore.
2. But, I encounter x2 difference in time between CLI and Studio, is it expected?

macOS:
CLI: 6 sec
Studio: 14 sec

Windows:
CLI: 1:40
Studio: 2:50

fredrikekelund committed:
Refine download-node-binary script

- ESM instead of CommonJS
- Use npm modules instead of shell commands to extract archives
- Linux isn't supported
- fetch API instead of https module

fredrikekelund committed:
Use bundled node binary in ./bin scripts

fredrikekelund committed:
Flip the logic for distinguishing node environments

fredrikekelund committed:
process.env is already the default

fredrikekelund committed:
Use system-level node binary in development

fredrikekelund committed:
Use Electron node binary in unit tests

fredrikekelund reviewed (approved):
Great initiative! I'm very optimistic about the performance impact of this change.

I made a couple of changes to this PR:
- Refined `scripts/download-node-binary.mjs` in 17e2ead
- Updated `./bin` scripts to use the bundled node binary in 4e77d33 (meaning Electron isn't involved when users run `studio` from the terminal, either)
- Use the system-level node binary when forking CLI processes during development in f0b9873 (this addresses the performance problems @nightnei noted in https://github.com/Automattic/studio/pull/2303#pullrequestreview-3611541580)

With those changes in place, this PR should be good to land 👍

gavande1 commented:
> But, I encounter x2 difference in time between CLI and Studio, is it expected?

@bcotrim @fredrikekelund I am curious, Did we addressed this issue?

fredrikekelund commented:
We did! The difference @nightnei noticed came from the fact that we previously used the Electron node runtime to fork CLI processes. In f0b9873, I made it so the system-level node binary is used instead, which means performance will be the same as when running the CLI directly, and we more closely match the production build behavior (which is to use the newly bundled node runtime).

wojtekn commented on scripts/download-node-binary.mjs:
```diff
@@ -0,0 +1,166 @@
+		const version = fs.readFileSync( nvmrcPath, 'utf-8' ).trim();
+		return version.startsWith( 'v' ) ? version : `v${ version }`;
+	}
+	console.log( `.nvmrc not found, using fallback version ${ LTS_FALLBACK }` );
```
Would it be safer to remove fallback and fail the job/build if there is no version in `.nvmrc`?

bcotrim replied:
We can even consider have a separate version for the bundled Node.
Keeping the Node version in sync with the recommended version from Playground CLI should be the best approach. cc @fredrikekelund
